From aad17c44dfe0c261305bf6d9f6392cfb36350ae1 Mon Sep 17 00:00:00 2001
From: "Yue, VincentX" <vincentx.yue@intel.com>
Date: Wed, 4 Apr 2018 16:15:56 +0800
Subject: [PATCH 1/2] Add permission and flags in Car Dialer

Add WAKE_LOCK permission in manifest and related flags in TelecomActivity
so that incoming call can be shown in front of the Keyguard.

Test:
1.Pair and Connect BXT-IVI to reference phone via BT
2.Set Security Pattern(Setting->Security->Screen Lock->Pattern)
3.Lock BXT-IVI
4.Make a call from Phone1 to BXT-IVI (DUT is in locked state)

Change-Id: I66f220fe54a47f2a92eca3cbcde44120351a7457
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-56843
Signed-off-by: robinz1x <robinx.zhang@intel.com>
Signed-off-by: Yue, VincentX <vincentx.yue@intel.com>
---
 AndroidManifest.xml                             |  1 +
 src/com/android/car/dialer/TelecomActivity.java | 19 +++++++++++++++++++
 2 files changed, 20 insertions(+)

diff --git a/AndroidManifest.xml b/AndroidManifest.xml
index d420c7e..571ab9b 100644
--- a/AndroidManifest.xml
+++ b/AndroidManifest.xml
@@ -19,6 +19,7 @@
 
     <uses-sdk android:minSdkVersion="24" android:targetSdkVersion='24'/>
 
+    <uses-permission android:name="android.permission.WAKE_LOCK"/>
     <uses-permission android:name="android.permission.BLUETOOTH"/>
     <uses-permission android:name="android.permission.READ_CALL_LOG"/>
     <uses-permission android:name="android.permission.READ_CONTACTS"/>
diff --git a/src/com/android/car/dialer/TelecomActivity.java b/src/com/android/car/dialer/TelecomActivity.java
index 06d74da..7df499e 100644
--- a/src/com/android/car/dialer/TelecomActivity.java
+++ b/src/com/android/car/dialer/TelecomActivity.java
@@ -29,6 +29,7 @@ import android.telecom.Call;
 import android.telephony.PhoneNumberUtils;
 import android.util.Log;
 import android.view.View;
+import android.view.WindowManager;
 
 import com.android.car.app.CarDrawerActivity;
 import com.android.car.app.CarDrawerAdapter;
@@ -198,6 +199,21 @@ public class TelecomActivity extends CarDrawerActivity implements
         setIntent(null);
     }
 
+    private void addDimissKeyguardFlag(){
+        // set these flags so this activity will stay in front of the keyguard
+        getWindow().addFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
+                | WindowManager.LayoutParams.FLAG_DISMISS_KEYGUARD
+                | WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
+                | WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON);
+    }
+
+    private void clearDimissKeyguardFlag(){
+        // clear these flags so this activity will stay in front of the keyguard
+        getWindow().clearFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
+                | WindowManager.LayoutParams.FLAG_DISMISS_KEYGUARD
+                | WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
+                | WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON);
+    }
     /**
      * Updates the content fragment of this Activity based on the state of the application.
      */
@@ -263,6 +279,7 @@ public class TelecomActivity extends CarDrawerActivity implements
             closeDrawer();
             return;
         }
+        addDimissKeyguardFlag();
 
         Fragment fragment = OngoingCallFragment.newInstance(mUiCallManager, mUiBluetoothMonitor);
         setContentFragmentWithFadeAnimation(fragment);
@@ -407,6 +424,8 @@ public class TelecomActivity extends CarDrawerActivity implements
                 Log.d(TAG, "onCallRemoved");
             }
             updateCurrentFragment();
+            clearDimissKeyguardFlag();
+
         }
 
         @Override
-- 
1.9.1

